### ISSUE TITLE
Is there a supported way to enable optional DFL and NMS behavior in YOLO26?

### ISSUE THREAD

**Issue by arda92a** [ultralytics/ultralytics#23514]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello Ultralytics team,

Thank you for the work on YOLO26 and the transition toward fully end-to-end detection. The removal of NMS and simplification of the regression pipeline are very interesting design changes.

I had a few clarification questions regarding flexibility and experimentation:

1) DFL Usage
--------------------------------
Is there any supported or recommended way to enable Distribution Focal Loss behavior in YOLO26?

More specifically:
- Is adjusting regression-related configuration parameters intended to allow experimentation with DFL-like behavior?
- Or is YOLO26 fundamentally designed to operate without DFL and such experimentation is discouraged?

2) Optional NMS Integration
--------------------------------
Since YOLO26 is designed to be NMS-free, I was wondering whether there is any supported or recommended way to:

- Apply optional NMS during inference, or
- Export models with integrated NMS similar to earlier YOLO versions.

In some dense object scenarios, additional filtering can still be useful, so I wanted to understand whether this is intentionally unsupported or simply left to user-side post-processing.

Thanks again for your work and for any clarification you can provide.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @arda92a, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Thanks for the thoughtful questions around YOLO26 design flexibility (DFL + optional NMS) 🙌 This is an automated response, and an Ultralytics engineer will also assist soon to clarify what’s supported vs. expected to be handled externally.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by onuralpszr** [comment]
@arda92a we have option to disable end2end and using classic NMS inference and also depend on export we also include those options as well

https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/default.yaml#L59

Also when you try export while end2end enable for example and enable nms=true, nms option automatically will set to false unless you set end2end false to enable it. We have old options also still usable.

**Comment by onuralpszr** [comment]
@arda92a Also for DFL Yes, adjusting regression parameters allows for experimentation. Despite the default setting, the underlying Detect head and loss classes (v8DetectionLoss) are fully capable of supporting DFL. If you change reg_max to a value greater than 1 (e.g., the standard 16 used in YOLOv8), the code will automatically instantiate the DFL module and generic DFL loss logic. This allows you to experiment with DFL behavior in the YOLO26 architecture, even if it wasn't the primary design choice for this specific version.

```python
from ultralytics import YOLO

# Load model, this will load the structure
model = YOLO("yolo26n.yaml") 

# Modify the reg_max in the model's yaml config before building
model.model.args["reg_max"] = 16

# Train the model with this new configuration
model.train(data="coco8.yaml", epochs=10)
```

**Comment by Y-T-G** [comment]
You can change the model config and retrain to use DFL and only support NMS based inference:

https://github.com/ultralytics/ultralytics/blob/dc22aad1e5c985dc423b4418ca805c16d33936ac/ultralytics/cfg/models/26/yolo26.yaml#L9-L10

`reg_max` needs to be set to 16 to enable DFL and `end2end` is set to `False` to make it like YOLO11.

However, NMS-free models support both NMS based and NMS-free inference like mentioned by @onuralpszr

**Comment by arda92a** [comment]
Thank you very much for the detailed clarification and examples. This was extremely helpful in understanding the flexibility of YOLO26 and how the architecture still supports both classic and end-to-end configurations.

**Comment by glenn-jocher** [comment]
Glad that clarifies it — one extra practical note: if you want to tune filtering in dense scenes (i.e., use `iou`), run inference with `end2end=False` (otherwise `iou` won’t apply) as described in the [Predict mode docs](https://docs.ultralytics.com/modes/predict/), and if you switch `reg_max` to re-enable DFL you’ll need to retrain since the head shape changes (see the `Detect(nc, reg_max, end2end, ...)` signature in the [Detect head reference](https://docs.ultralytics.com/reference/nn/modules/head/)).
